### ISSUE TITLE
OAuth2 and Office 365/Hosted Exchange

### ISSUE THREAD

**Issue by AdhocAdam** [AdhocAdam/smletsexchangeconnector#171]
**UPDATE**: [Basic Authentication to be disabled in second half of 2021 for those actively using it](https://techcommunity.microsoft.com/t5/exchange-team-blog/basic-authentication-and-exchange-online-april-2020-update/ba-p/1275508).

~~With the upcoming [deprecation of Basic Authentication to Office 365 through Exchange Web Services API on October 13th, 2020](https://techcommunity.microsoft.com/t5/exchange-team-blog/upcoming-changes-to-exchange-web-services-ews-api-for-office-365/ba-p/608055)~~. It's a priority for those running their Service Manager inbox in 365 that the connector continues to work.

This request sees to enable the connector to handle on-premise and hosted Exchange environments by introducing support for OAuth2 tokens as a means to authenticate to Exchange.

High level steps include:
- Registering a new Azure Active Directory Application
    - Assigning the Application API Permissions _(one or the other)_
        - Microsoft Graph (i.e. _Delegated_ Mail.ReadWrite and Mail.Send)
        - Exchange (i.e. _Delegated_ EWS.AccessAsUser.All)
- One time configuration for the workflow account to consent to use the Azure AD Application
- Calling the Application to retrieve mail for the workflow account _(one or the other)_
   - **Microsoft Graph**
        - returns messages as odata objects
        - removes the dependency on the EWS.dll in favor of pure PowerShell, Invoke-RestMethod style calls
        - requires re-factoring of the connector and the mail/calendar objects in the primary loop as well as the function that sends mail as the workflow account
   - **Exchange**
        - requires upgrade to at least v2.0 of the EWS.dll to work with OAuth tokens as credentials

**Comment by AdhocAdam** [comment]
To be refined -

1. Register an App in Azure Active Directory
![01](https://user-images.githubusercontent.com/6636040/75416766-dd858100-58f4-11ea-82a1-c97df9a32f20.PNG)

2. Take note of the App information
![03](https://user-images.githubusercontent.com/6636040/75416805-faba4f80-58f4-11ea-8418-bc111bba8c3a.PNG)

3. Generate a secret for the app
![04](https://user-images.githubusercontent.com/6636040/75416807-fb52e600-58f4-11ea-84b8-d82af4521507.PNG)

4. Choose the APIs the app can leverage
![Untitled-1](https://user-images.githubusercontent.com/6636040/75416872-2a695780-58f5-11ea-89d4-aac0e42f263d.png)

5. Workflow account to one time consent to the app's permissions (i.e. read mailbox)
6. Configure the connector to use workflow account + app secret to process the inbox

**Comment by AdhocAdam** [comment]
**Assuming EWS is used**.
Per the Microsoft documentation [Authenticate an EWS application by using OAuth](https://docs.microsoft.com/en-us/exchange/client-developer/exchange-web-services/how-to-authenticate-an-ews-application-by-using-oauth) and then translating the C# into PowerShell:

```powershell
$tenantName = "" #this is your Azure Tenant Name e.g. mydomain.onmicrosoft.com
$clientId = "" #this is the application/client id of the Azure Active Directory app from above
$clientSecret = "" #this is the application secret generated above
$username = "" #the workflow account's username
$password = "" #the workflow account's password

$tokenReqBody = @{
    Grant_Type    = "Password"
    Client_Id     = $clientID
    Client_Secret = $clientSecret
    Username      = $username
    Password      = $password
    Scope         = "https://outlook.office.com/EWS.AccessAsUser.All"
}
$tokenReqResponse = Invoke-RestMethod -Uri "https://login.microsoftonline.com/$TenantName/oauth2/v2.0/token" -Method "POST" -Body $tokenReqBody 

#define Exchange assembly and connect to Office 365 EWS
$exchangeEWSAPIPath = "C:\Program Files\Microsoft\Exchange\Web Services\2.0\Microsoft.Exchange.WebServices.dll"
$exchangeEndpoint = "https://outlook.office365.com/EWS/Exchange.asmx"
[void] [Reflection.Assembly]::LoadFile("$exchangeEWSAPIPath")
$exchangeService = New-Object Microsoft.Exchange.WebServices.Data.ExchangeService
$exchangeService.Url = [System.Uri]$ExchangeEndpoint
$exchangeService.Credentials = [Microsoft.Exchange.WebServices.Data.OAuthCredentials]($tokenReqResponse.access_token)

#define search parameters, search on the defined classes and get messages that are older than the current time
$inboxFolderName = [Microsoft.Exchange.WebServices.Data.WellKnownFolderName]::Inbox
$inboxFolder = [Microsoft.Exchange.WebServices.Data.Folder]::Bind($exchangeService,$inboxFolderName)
```

At this point the user is authenticated and is able to process the inbox through Exchange Web Services in Office 365 using OAuth tokens.

**Comment by AdhocAdam** [comment]
**Assuming the Microsoft Graph API is used.**
The mail is retrieved through an Invoke-RestMethod call to the Graph.

```powershell
$tenantName = "" #this is your Azure Tenant Name e.g. mydomain.onmicrosoft.com
$clientId = "" #this is the application/client id of the Azure Active Directory app from above
$clientSecret = "" #this is the application secret generated above
$username = "" #the workflow account's username
$password = "" #the workflow account's password

$tokenReqBody = @{
    Grant_Type    = "Password"
    Client_Id     = $clientID
    Client_Secret = $clientSecret
    Username      = $username
    Password      = $password
    Scope         = "https://graph.microsoft.com/Mail.Read"
} 

#get the token to use with the Graph
$tokenReqResponse = Invoke-RestMethod -Uri "https://login.microsoftonline.com/$TenantName/oauth2/v2.0/token" -Method "POST" -Body $tokenReqBody 

### make a call and ask for the Message Class (e.g. IPM.Note) property since it isn't included in default call: ($filter = Id eq String 0x001a)
$apiUrl = "https://graph.microsoft.com/v1.0/me/mailFolders/Inbox/messages/?`$expand=SingleValueExtendedProperties(`$filter=(Id%20eq%20'String%200x001a'))"
$data = Invoke-RestMethod -Headers @{Authorization = "Bearer $($tokenReqResponse.access_token)"} -Uri $apiUrl -Method "GET"
$inbox = $data.value | select from, toRecipients, ccRecipients, subject, sentDateTime, receivedDateTime, id, conversationid, hasattachments, @{Name = 'ItemClass'; Expression = {$_.singleValueExtendedProperties | select value -ExpandProperty value }}

```

At this point the user is authenticated and is able to process the inbox through the Microsoft Graph in Office 365 using OAuth tokens.

**Comment by AdhocAdam** [comment]
The recent release of [Exchange Connector 4.0](https://www.microsoft.com/en-us/download/details.aspx?id=101579) has now introduced OAuth support for Exchange Online. In an effort to continue to maintain a low barrier to entry to this connector as well as continued support for those currently using it in 365. A similar authentication method will be followed wherein:

- An Azure Application is deployed into your own tenant (few changes to previously outlined screenshots above)
    - Redirect URI to change from "Web" to "Public client/native (mobile & desktop)
    - Redirect URI to be set to "ietf:wg:oauth:2.0.oob"
    - Do not generate a Secret for the App
- Flexibility around Delegated "EWS.AccessAsUser.All" permission. Administrators can choose to either:
    - "Grant admin consent for org" for the recently registered app and it's EWS permission
    - Run the one time Consent script (seen above) as the SCSM workflow account to explicitly grant EWS access for Service Manager's workflow account

**Comment by AdhocAdam** [comment]
OAuth functionality to be available with #206

**Comment by AdhocAdam** [comment]
The location of EWS has moved within the Azure Portal.

![graphAPI](https://user-images.githubusercontent.com/6636040/113810290-8674f100-972f-11eb-8121-f261f488e15f.PNG)
